pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/retention/RetentionManager.java

```diff
@@ -165,6 +187,95 @@ private void manageRetentionForRealtimeTable(String realtimeTableName, Retention
     }
   }
 
+  @VisibleForTesting
+  void manageRetentionForHybridTable(TableConfig realtimeTableConfig) {
+    LOGGER.info("Managing retention for hybrid table: {}", realtimeTableConfig.getTableName());
+    List<String> segmentsToDelete = new ArrayList<>();
+    String realtimeTableName = realtimeTableConfig.getTableName();
+    String rawTableName = TableNameBuilder.extractRawTableName(realtimeTableName);
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    try {
+      TableConfig offlineTableConfig = _pinotHelixResourceManager.getOfflineTableConfig(offlineTableName);
+      ZkHelixPropertyStore<ZNRecord> propertyStore = _pinotHelixResourceManager.getPropertyStore();
+      Schema schema = ZKMetadataProvider.getTableSchema(propertyStore, offlineTableName);
+      if (schema == null) {
+        throw new RuntimeException("Failed to get schema for table: " + offlineTableName);
+      }
+      String timeColumn = null;
+      SegmentsValidationAndRetentionConfig validationConfig = offlineTableConfig.getValidationConfig();
+      if (validationConfig != null) {
+        timeColumn = validationConfig.getTimeColumnName();
+      }
+      if (StringUtils.isEmpty(timeColumn)) {
+        throw new RuntimeException("TimeColumn is null or empty for table: " + offlineTableName);
+      }
+      DateTimeFieldSpec dateTimeSpec = schema.getSpecForTimeColumn(timeColumn);
+      if (dateTimeSpec == null) {
+        throw new RuntimeException(String.format("Failed to get DateTimeFieldSpec for time column: %s of table: %s",
+            timeColumn, offlineTableName));
+      }
+      DateTimeFormatSpec timeFormatSpec = dateTimeSpec.getFormatSpec();
+      if (timeFormatSpec.getColumnUnit() == null) {
+        throw new RuntimeException(String.format("Time unit must be configured in the field spec for time column: %s of"
+            + " table: %s", timeColumn, offlineTableName));
+      }
+      List<SegmentZKMetadata> offlineSegmentsZKMetadata =
```

Comment:
Reading all ZK metadata for OFFLINE table can be very expensive. Time boundary info can be read from broker API. See `PinotBrokerDebug.getTimeBoundary()`